classes/Deck.js

```diff
@@ -1,22 +1,187 @@
-import { PlayingCard } from './PlayingCard'
+import { SUITES, RANKS } from '../constants/playing-card'
 
 class Deck {
-  constructor(suites, ranks) {
-    const cards = []
-    for (const suite of suites) {
-      for (const rank of ranks) {
-        cards.push(new PlayingCard(suite, rank))
+  constructor(cards = []) {
+    this.cards = cards
+  }
+
+  validateCanDrawCard(cardPosition) {
+    if ((0 >= cardPosition) || (cardPosition > this.size())) {
+      throw new Error(`RangeError: Cannot remove card from position (${cardPosition})`)
+    }
+  }
+
+  validateCanInsertCard(cardPosition) {
+    if ((0 >= cardPosition) || (cardPosition > this.size() + 1)) {
+      throw new Error(`RangeError: Cannot insert card to position ${cardPosition}`)
+    }
+  }
+
+  validateNotEmpty() {
+    if (this.isEmpty()) {
+      throw new Error(`IllegalOperation: Cannot draw card, deck is empty`)
+    }
+  }
+
+  validateNotFull() {
+    if (this.isFull()) {
+      throw new Error(`IllegalOperation: Cannot insert card, deck is full`)
+    }
+  }
+
+  print() {
+    this.cards.forEach(console.log)
+  }
+
+  drawCard(cardPosition) {
+    this.validateNotEmpty()
+    this.validateCanDrawCard(cardPosition)
+
+    return this.cards.splice(cardPosition - 1, 1)[0]
+  }
+
+  drawTopCard() {
+    return this.drawCard(this.size())
+  }
+  
+  drawBottomCard() {
+    return this.drawCard(1)
+  }
+
+  insertCard(card, cardPosition) {
+    this.validateNotFull()
+    this.validateCanInsertCard(cardPosition)
+
+    const { bottom, top } = this.split(cardPosition - 1)
+    bottom.cards.push(card)
+    
+    this.stack(bottom.stack(top))
+  }
+
+  insertCardTop(card) {
+    this.insertCard(card, this.size() + 1)
+  }
+
+  insertCardBottom(card) {
+    this.insertCard(card, 1)
+  }
+
+  isEmpty() {
+    return this.size() === 0
+  }
+
+  isFull() {
+    throw new Error(`Missing implementation for this function`)
+  }
+
+  size() {
+    return this.cards.length
+  }
+
+  stack(deck) {
+    this.cards = this.cards.concat(deck.cards)
+
+    return this
+  }
+
+  split(position) {
+    return { 
+      bottom: new Deck(this.cards.splice(0, position)),
+      top: new Deck(this.cards.splice(0, this.cards.length))
+    }
+  }
```

Comment:
The class methods should all be consistent about the card positions. The class methods shouldn't receive the already "corrected" position for the underlying data structure. You should only do that "correction" when using the Deck's data structure API, and therefore in insertCard, for example, I suggest passing `position` to the call to `split`, **not** `position - 1`, and then here pass `position - 1` to `splice`.